File: Striver/DP/Q23.java
package Striver.DP;

public class Q23 {
    // Unbounded Knapsack Problem

    // recursive solution - Time: O(2^n) Space: O(n)
    public static int unboundedKnapsack(int[] wt, int[] val, int W, int n) {
        if(n == 0){
            return (int)((W/wt[n])*val[n]);
        }
        int notTake = unboundedKnapsack(wt, val, W, n-1);
        int take = 0;
        if(W >= wt[n]){
            take = unboundedKnapsack(wt, val, W-wt[n], n) + val[n];
        }
        return Math.max(notTake, take);
    }

    // memoization - Time: O(n*W) Space: O(n*W)
    public static int unboundedKnapsackMemo(int[] wt, int[] val, int W, int n, int[][] dp) {
        if(n == 0){
            return (int)((W/wt[n])*val[n]);
        }
        if(dp[n][W] != -1){
            return dp[n][W];
        }
        int notTake = unboundedKnapsackMemo(wt, val, W, n-1, dp);
        int take = 0;
        if(W >= wt[n]){
            take = unboundedKnapsackMemo(wt, val, W-wt[n], n, dp) + val[n];
        }
        return dp[n][W] = Math.max(notTake, take);
    }

    // tabulation - Time: O(n*W) Space: O(n*W)
    public static int unboundedKnapsackTab(int[] wt, int[] val, int W, int n) {
        int[][] dp = new int[n][W+1];

        for(int i = 0; i <= W; i++){
            dp[0][i] = (int)((i/wt[0])*val[0]);
        }

        for(int i = 1; i < n; i++){
            for(int j = 0; j <= W; j++){
                int notTake = dp[i-1][j];
                int take = 0;
                if(j >= wt[i]){
                    take = dp[i][j-wt[i]] + val[i];
                }
                dp[i][j] = Math.max(notTake, take);
            }
        }

        return dp[n-1][W];
    }

    // space optimization - Time: O(n*W) Space: O(W)
    public static int unboundedKnapsackSpaceOpt(int[] wt, int[] val, int W, int n) {
        int[] dp = new int[W+1];
        int[] curr = new int[W+1];

        for(int i = 0; i <= W; i++){
            dp[i] = (int)((i/wt[0])*val[0]);
        }

        for(int i = 1; i < n; i++){
            for(int j = 0; j <= W; j++){
                int notTake = dp[j];
                int take = 0;
                if(j >= wt[i]){
                    take = curr[j-wt[i]] + val[i];
                }
                curr[j] = Math.max(notTake, take);
            }
            dp = curr;
            curr = new int[W+1];
        }

        return dp[W];
    }

    public static void main(String[] args) {
        int[] wt = {1, 3, 4, 5};
        int[] val = {10, 40, 50, 70};
        int W = 8;
        int n = wt.length-1;

        System.out.println(unboundedKnapsack(wt, val, W, n));
        System.out.println(unboundedKnapsackMemo(wt, val, W, n, new int[n+1][W+1]));
        System.out.println(unboundedKnapsackTab(wt, val, W, n));
        System.out.println(unboundedKnapsackSpaceOpt(wt, val, W, n));
    }
}

File: Striver/DP/Q23.py
# Unbounded KnapSack

# recursive solution
def unboundedKnapSack(W, wt, val, n):
    if n == 0:
        return (W//wt[n])*val[n]
    
    notTake = unboundedKnapSack(W, wt, val, n-1)
    take = 0
    if W >= wt[n]:
        take = unboundedKnapSack(W-wt[n], wt, val, n) + val[n]
    return max(notTake, take)

# memoization
def unboundedKnapSackMemo(W, wt, val, n, dp):
    if n == 0:
        return (W//wt[n])*val[n]
    
    if dp[n][W] != -1:
        return dp[n][W]
    
    notTake = unboundedKnapSackMemo(W, wt, val, n-1, dp)
    take = 0
    if W >= wt[n]:
        take = unboundedKnapSackMemo(W-wt[n], wt, val, n, dp) + val[n]
    dp[n][W] = max(notTake, take)
    return dp[n][W]

# tabulation
def unboundedKnapSackTab(W, wt, val, n):
    dp = [[0 for i in range(W+1)] for j in range(n)]

    for i in range(W+1):
        dp[0][i] = (i//wt[0])*val[0]
    
    for i in range(1, n):
        for j in range(1, W+1):
            notTake = dp[i-1][j]
            take = 0
            if j >= wt[i]:
                take = dp[i][j-wt[i]] + val[i]
            dp[i][j] = max(notTake, take)

    return dp[n-1][W]

# space optimization
def unboundedKnapSackSpaceOpt(W, wt, val, n):
    dp = [0 for i in range(W+1)]
    curr = [0 for i in range(W+1)]

    for i in range(W+1):
        dp[i] = (i//wt[0])*val[0]

    for i in range(1, n):
        for j in range(1, W+1):
            notTake = dp[j]
            take = 0
            if j >= wt[i]:
                take = curr[j-wt[i]] + val[i]
            curr[j] = max(notTake, take)
        dp = curr
        curr = [0 for i in range(W+1)]

    return dp[W]

vals = [10, 40, 50, 70]
wts = [1, 3, 4, 5]
W = 8
n = len(vals)
dp = [[-1 for i in range(W+1)] for j in range(n)]
print(unboundedKnapSack(W, wts, vals, n-1))
print(unboundedKnapSackMemo(W, wts, vals, n-1, dp))
print(unboundedKnapSackTab(W, wts, vals, n))
print(unboundedKnapSackSpaceOpt(W, wts, vals, n))
